<a href="https://colab.research.google.com/github/Sru-j/daily_lab_activities/blob/main/data_inspect.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# import packages
import pandas as pd

## 1) Load the messy dataset (with headers)

In [2]:
# messy_employee_health.csv
url = "https://raw.githubusercontent.com/bkeerthana/MLC/refs/heads/main/messy_employee_health.csv"
df = pd.read_csv(url)
df

ID  Full Name   Age    GENDER Heart Rate  Height  \
0     1.0       Anu   22.0   Female     72 bpm  170 cm   
1     2.0     Rahul    NaN     MALE      85bpm  1.75 m   
2     3.0     Meera   29.0  female         NaN  175 cm   
3     4.0      Sidd   -5.0     Male     90 bpm   180cm   
4     5.0     Priya   31.0   FEMALE    100 BPM  1.60 m   
5     6.0     Kiran  120.0     male         ??  160 cm   
6     7.0       Anu   22.0   Female     72 bpm  170 cm   
7     8.0      Zara   27.0   FeMale     65 bpm     NaN   
8     9.0       NaN   24.0     male     75 bpm  175 cm   
9    10.0     Vijay   28.0      Mle     80 bpm  170 cm   
10    NaN       NaN    NaN      NaN        NaN     NaN   

                                   Address       Join Date       Notes  \
0            12 MG Road, Bengaluru, 560001      2024-01-05        good   
1              45 Park St, Kolkata, 700016      05/02/2024  naïve café   
2           77 Anna Salai, Chennai, 600002      2024/03/10         NaN   
3              1 Marine Dr, Mumbai, 400001  March 15, 2024           ?   
4   No. 9, Residency Rd, Bengaluru, 560025      2024-13-01         NaN   
5                      Somewhere, , 560002      2024-06-01     unknown   
6            12 MG Road, Bengaluru, 560001      2024-01-05        good   
7           8 Lake View, Hyderabad, 500001      2024-07-04           —   
8          15 Church St, Bengaluru, 560001             NaN          ok   
9               18 North St, Delhi, 110001      2024-09-01          ok   
10                                     NaN             NaN         NaN   

   Constant Column Irrelevant Column  Empty Column  
0             SAME                x1           NaN  
1             SAME                x2           NaN  
2             SAME                x3           NaN  
3             SAME                x4           NaN  
4             SAME                x5           NaN  
5             SAME                x6           NaN  
6             SAME                x1           NaN  
7             SAME                x7           NaN  
8             SAME                x8           NaN  
9             SAME                x9           NaN  
10             NaN               NaN           NaN

data inspection

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0     ID               10 non-null     float64
 1   Full Name          9 non-null      object 
 2   Age                9 non-null      float64
 3   GENDER             10 non-null     object 
 4   Heart Rate         9 non-null      object 
 5   Height             9 non-null      object 
 6   Address            10 non-null     object 
 7   Join Date          9 non-null      object 
 8   Notes              8 non-null      object 
 9   Constant Column    10 non-null     object 
 10  Irrelevant Column  10 non-null     object 
 11  Empty Column       0 non-null      float64
dtypes: float64(3), object(9)
memory usage: 1.2+ KB


In [4]:
df.describe()

ID         Age   Empty Column
count  10.00000    9.000000           0.0
mean    5.50000   33.111111           NaN
std     3.02765   34.309053           NaN
min     1.00000   -5.000000           NaN
25%     3.25000   22.000000           NaN
50%     5.50000   27.000000           NaN
75%     7.75000   29.000000           NaN
max    10.00000  120.000000           NaN

In [5]:
df.columns

Index(['  ID ', 'Full Name', 'Age ', 'GENDER', 'Heart Rate', 'Height',
       'Address', 'Join Date', 'Notes', 'Constant Column', 'Irrelevant Column',
       'Empty Column'],
      dtype='object')

In [6]:
df.isnull().sum()

ID                  1
Full Name             2
Age                   2
GENDER                1
Heart Rate            2
Height                2
Address               1
Join Date             2
Notes                 3
Constant Column       1
Irrelevant Column     1
Empty Column         11
dtype: int64

In [7]:
type(df)

pandas.core.frame.DataFrame

In [8]:
df.dtypes

ID                 float64
Full Name             object
Age                  float64
GENDER                object
Heart Rate            object
Height                object
Address               object
Join Date             object
Notes                 object
Constant Column       object
Irrelevant Column     object
Empty Column         float64
dtype: object

In [9]:
df.shape

(11, 12)

In [10]:
df.size

132

In [11]:
df.duplicated().sum()

np.int64(0)

changing column names

In [12]:
df.columns = df.columns.str.strip().str.lower().str.replace(' ','_')
df.columns

Index(['id', 'full_name', 'age', 'gender', 'heart_rate', 'height', 'address',
       'join_date', 'notes', 'constant_column', 'irrelevant_column',
       'empty_column'],
      dtype='object')

data cleaning

In [13]:
df.shape

(11, 12)

In [14]:
#drop the empty rows
df = df.dropna(axis=0, how='all')

In [15]:
df.shape

(10, 12)

In [16]:
#drop the empty columns
df = df.dropna(axis=1, how='all')

In [17]:
df.shape

(10, 11)

In [18]:
df['join_date'][4] = '2025-01-01'

/tmp/ipython-input-3051752583.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df['join_date'][4] = '2025-01-01'


changing data types

In [19]:
df['id'] = df['id'].astype('Int64')

In [20]:
df['age'] = df['age'].astype('Int64')

In [21]:
df['gender'] = df['gender'].astype('category')

In [22]:
df['gender'].str.lower().str.strip()

0    female
1      male
2    female
3      male
4    female
5      male
6    female
7    female
8      male
9       mle
Name: gender, dtype: object

In [23]:
df['join_date'] = df['join_date'].astype('datetime64[ns]')

In [24]:
df[df['age'] < 0]

id full_name  age gender heart_rate height                      address  \
3   4      Sidd   -5   Male     90 bpm  180cm  1 Marine Dr, Mumbai, 400001   

   join_date notes constant_column irrelevant_column  
3 2024-03-15     ?            SAME                x4

In [25]:
df[df['age'] > 100]

id full_name  age gender heart_rate  height              address  \
5   6     Kiran  120   male         ??  160 cm  Somewhere, , 560002   

   join_date    notes constant_column irrelevant_column  
5 2024-06-01  unknown            SAME                x6

In [26]:
#changing the age greater than 100 to 100
df.loc[df['age'] > 100, 'age'] = 100

In [27]:
#change the negative value of age to its average value
df['age'] = df['age'].abs()

In [28]:
df['age']

0      22
1    <NA>
2      29
3       5
4      31
5     100
6      22
7      27
8      24
9      28
Name: age, dtype: Int64

In [29]:
#expant is to tell that make a new column for each split
#n is number of columns to make starting from 0
df[['street','city','pincode']] = df['address'].str.split(',',n=2, expand=True)
df

id full_name   age   gender heart_rate  height  \
0   1       Anu    22   Female     72 bpm  170 cm   
1   2     Rahul  <NA>     MALE      85bpm  1.75 m   
2   3     Meera    29  female         NaN  175 cm   
3   4      Sidd     5     Male     90 bpm   180cm   
4   5     Priya    31   FEMALE    100 BPM  1.60 m   
5   6     Kiran   100     male         ??  160 cm   
6   7       Anu    22   Female     72 bpm  170 cm   
7   8      Zara    27   FeMale     65 bpm     NaN   
8   9       NaN    24     male     75 bpm  175 cm   
9  10     Vijay    28      Mle     80 bpm  170 cm   

                                  address  join_date       notes  \
0           12 MG Road, Bengaluru, 560001 2024-01-05        good   
1             45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2          77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3             1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4  No. 9, Residency Rd, Bengaluru, 560025 2025-01-01         NaN   
5                     Somewhere, , 560002 2024-06-01     unknown   
6           12 MG Road, Bengaluru, 560001 2024-01-05        good   
7          8 Lake View, Hyderabad, 500001 2024-07-04           —   
8         15 Church St, Bengaluru, 560001        NaT          ok   
9              18 North St, Delhi, 110001 2024-09-01          ok   

  constant_column irrelevant_column         street           city  \
0            SAME                x1     12 MG Road      Bengaluru   
1            SAME                x2     45 Park St        Kolkata   
2            SAME                x3  77 Anna Salai        Chennai   
3            SAME                x4    1 Marine Dr         Mumbai   
4            SAME                x5          No. 9   Residency Rd   
5            SAME                x6      Somewhere                  
6            SAME                x1     12 MG Road      Bengaluru   
7            SAME                x7    8 Lake View      Hyderabad   
8            SAME                x8   15 Church St      Bengaluru   
9            SAME                x9    18 North St          Delhi   

              pincode  
0              560001  
1              700016  
2              600002  
3              400001  
4   Bengaluru, 560025  
5              560002  
6              560001  
7              500001  
8              560001  
9              110001

In [30]:
df.drop(columns=['address'], inplace=True)
df

id full_name   age   gender heart_rate  height  join_date       notes  \
0   1       Anu    22   Female     72 bpm  170 cm 2024-01-05        good   
1   2     Rahul  <NA>     MALE      85bpm  1.75 m 2024-05-02  naïve café   
2   3     Meera    29  female         NaN  175 cm 2024-03-10         NaN   
3   4      Sidd     5     Male     90 bpm   180cm 2024-03-15           ?   
4   5     Priya    31   FEMALE    100 BPM  1.60 m 2025-01-01         NaN   
5   6     Kiran   100     male         ??  160 cm 2024-06-01     unknown   
6   7       Anu    22   Female     72 bpm  170 cm 2024-01-05        good   
7   8      Zara    27   FeMale     65 bpm     NaN 2024-07-04           —   
8   9       NaN    24     male     75 bpm  175 cm        NaT          ok   
9  10     Vijay    28      Mle     80 bpm  170 cm 2024-09-01          ok   

  constant_column irrelevant_column         street           city  \
0            SAME                x1     12 MG Road      Bengaluru   
1            SAME                x2     45 Park St        Kolkata   
2            SAME                x3  77 Anna Salai        Chennai   
3            SAME                x4    1 Marine Dr         Mumbai   
4            SAME                x5          No. 9   Residency Rd   
5            SAME                x6      Somewhere                  
6            SAME                x1     12 MG Road      Bengaluru   
7            SAME                x7    8 Lake View      Hyderabad   
8            SAME                x8   15 Church St      Bengaluru   
9            SAME                x9    18 North St          Delhi   

              pincode  
0              560001  
1              700016  
2              600002  
3              400001  
4   Bengaluru, 560025  
5              560002  
6              560001  
7              500001  
8              560001  
9              110001

In [31]:
df.dtypes

id                            Int64
full_name                    object
age                           Int64
gender                     category
heart_rate                   object
height                       object
join_date            datetime64[ns]
notes                        object
constant_column              object
irrelevant_column            object
street                       object
city                         object
pincode                      object
dtype: object

In [32]:
def convert_height(value):
    val = str(value).lower()
    if 'cm' in val:
        return float(val.replace('cm', '').strip())
    elif 'm' in val:
        # Convert meters to cm
        return float(val.replace('m', '').strip()) * 100
    else:
        # Handle cases with no unit, assuming cm
        return float(val)

In [33]:
df['height'] = df['height'].apply(convert_height)
df['height']

0    170.0
1    175.0
2    175.0
3    180.0
4    160.0
5    160.0
6    170.0
7      NaN
8    175.0
9    170.0
Name: height, dtype: float64

In [34]:
df['height'] = df['height'].fillna(0)
df['height']

0    170.0
1    175.0
2    175.0
3    180.0
4    160.0
5    160.0
6    170.0
7      0.0
8    175.0
9    170.0
Name: height, dtype: float64

In [35]:
#replace 0 in height to its mean value
df['height'] = df['height'].replace(0, df['height'].mean())
df['height']

0    170.0
1    175.0
2    175.0
3    180.0
4    160.0
5    160.0
6    170.0
7    153.5
8    175.0
9    170.0
Name: height, dtype: float64

In [41]:
df['heart_rate'] = df['heart_rate'].astype('Int64')
df['heart_rate'] = df['heart_rate'].fillna(0)
df['heart_rate']

0     72
1     85
2      0
3     90
4    100
5      0
6     72
7     65
8     75
9     80
Name: heart_rate, dtype: Int64

In [42]:
df['heart_rate'] = df['heart_rate'].str.lower().str.replace('bpm','')
df['heart_rate'] = df['heart_rate'].str.replace('?','0')
df['heart_rate']

AttributeError: Can only use .str accessor with string values!

In [38]:
df['heart_rate'] = df['heart_rate'].replace(0, df['heart_rate'].mean())

TypeError: can only concatenate str (not "int") to str

## 2) Missing headers (use the no-header file)


In [ ]:
# check the first first records

## 3) Missing values
Compute missing counts and percentages. Then choose a strategy (drop, fill, impute).

In [ ]:
# drop the missing values

In [ ]:
# fill the missing values with average

In [ ]:
# impute the values

## 4) Empty rows and empty columns
Identify and remove completely empty rows and columns.

In [ ]:
# TODO: find empty rows

In [ ]:
# TODO: drop empty rows

In [ ]:
# TODO: find empty columns

In [ ]:
# TODO: drop empty columns

In [ ]:
# df_step.shape

## 5) Messy column names
Standardize: strip, lowercase, underscores, remove special chars.

In [ ]:
# TODO: clean column names




## 6) Multiple variables in one column (Heart Rate)
Extract numeric heart rate (bpm) as an integer. Handle values like `85bpm`, `100 BPM`, `??`.

In [ ]:
# TODO: normalize heart_rate strings then extract digits

In [ ]:

# df_step['heart_rate'] = ...
# df_step['heart_rate_bpm'] = ...

# df_step[['heart_rate', 'heart_rate_bpm']].head(10)

## 7) Inconsistent units (Height)
Convert height to meters (float). Handle `170 cm`, `1.75 m`, `180cm`.

In [ ]:
# TODO: create height_m
# df_step['height_m'] = ...

# df_step[['height', 'height_m']].head(10)

## 8) Capitalization + typos in categorical values (Gender)
Standardize gender values and fix typos like `Mle`.

In [ ]:
# TODO: normalize gender
# df_step['gender'] = ...

# df_step['gender'].value_counts(dropna=False)

## 9) Data types (Age, IDs)
Convert `id` to integer (nullable ok), `age` to numeric; identify invalid values (e.g., negative age).

In [ ]:
# TODO: convert types
# df_step['id'] = ...
# df_step['age'] = ...

# TODO: find invalid ages
# df_step[(df_step['age'] < 0) | (df_step['age'] > 100)]

## 10) Addresses
Split address into street, city, and pincode. Handle missing parts safely.

In [ ]:
# TODO: split address by comma into 3 parts
# addr = df_step['address'].str.split(',', n=2, expand=True)
# df_step['street'] = ...
# df_step['city'] = ...
# df_step['pincode'] = ...

# df_step[['address','street','city','pincode']].head(10)

## 11) Dates
Parse join_date with mixed formats and coerce invalid dates to NaT.

In [ ]:
# TODO: parse dates
# df_step['join_date'] = pd.to_datetime(df_step['join_date'], errors='coerce', dayfirst=True)
# df_step[['join_date']].head(10)

## 12) Non-ASCII characters (Notes)
Remove non-ASCII characters OR normalize them.

In [ ]:
# TODO: remove non-ascii
# df_step['notes_clean'] = ...
# df_step[['notes','notes_clean']].head(10)

## 13) Unnecessary data
Remove duplicate rows, irrelevant columns, and uninformative columns.

In [ ]:
# TODO: duplicates
# df_step.duplicated().sum()
# df_step = df_step.drop_duplicates()

# TODO: drop irrelevant columns
# df_step = df_step.drop(columns=['irrelevant_column'], errors='ignore')

# TODO: drop uninformative columns (nunique == 1)
# nun = df_step.nunique(dropna=False)
# drop_cols = nun[nun == 1].index
# df_step = df_step.drop(columns=drop_cols)

# df_step.shape

## 14) Final cleaned dataset
Show summary and export as CSV.